File: DSA/cpp/Binary_Search.cpp
/* Binary Search Algorithm

Time Complexity:
The time complexity of the binary search algorithm is O(log n). The best-case time complexity would be O(1) when the central index would directly match the desired value. The worst-case scenario could be the values at either extremity of the list or values not in the list.

Space Complexity:
The space complexity of the binary search algorithm depends on the implementation of the algorithm. There are two ways of implementing it:
Iterative method - the space complexity would be O(1).
Recursive method - the space complexity would be O(log n).
*/

// Iterative Method -

#include <iostream>
using namespace std;

int binarySearch(int array[], int x, int low, int high) {

    // Repeat until the pointers low and high meet each other
    while (low <= high) {
        int mid = low + (high - low) / 2;

        if (array[mid] == x)
            return mid;

        if (array[mid] < x)
            low = mid + 1;

        else
            high = mid - 1;
    }

    return -1;
}

int main() {
    int array[] = {3, 4, 5, 6, 7, 8, 9};
    int x = 4;
    int n = sizeof(array) / sizeof(array[0]);
    int result = binarySearch(array, x, 0, n - 1);
    if (result == -1)
        cout << "Not found";
    else
        cout << "Element is found at index " << result;
}

// Recursive Method

#include <iostream>
using namespace std;

int binarySearch(int array[], int x, int low, int high) {
    if (high >= low) {
        int mid = low + (high - low) / 2;

        // If found at mid, then return it
        if (array[mid] == x)
            return mid;

        // Search the left half
        if (array[mid] > x)
            return binarySearch(array, x, low, mid - 1);

        // Search the right half
        return binarySearch(array, x, mid + 1, high);
    }

    return -1;
}

int main() {
    int array[] = {3, 4, 5, 6, 7, 8, 9};
    int x = 4;
    int n = sizeof(array) / sizeof(array[0]);
    int result = binarySearch(array, x, 0, n - 1);
    if (result == -1)
        cout << "Not found";
    else
        cout << "Element is found at index " << result;
}

File: DSA/Python/binary_search.py

# Recursive Binary Search Algorithm
# Binary Search is used to search an element in a sorted array(ascending)
# Time complexity = O (Log N)  where Log has the base 2

def binary_search(list, low, high, key):

    if high >= low:
        # Calculate Mid value,using [low +(high-low)/2] to deal with larger values
        mid = low+(high-low)//2

        # If element is present at the middle itself
        if list[mid] == key:
            return mid

        # If element is smaller than mid, then it can only
        # be present in left subarray
        elif list[mid] > key:
            return binary_search(list, low, mid - 1, key)

        # Else the element can only be present in right subarray
        else:
            return binary_search(list, mid + 1, high, key)

    else:
        # Element is not present in the array
        return -1


list = []

n = int(input("\nEnter Size of List : "))  # take size of list from user
print("\nEnter Values :")  # take values
for i in range(n):
    z = int(input())
    list.append(z)

k = int(input("\nEnter Element to be found : "))  # enter element to be found
list.sort()  # sort the list
print("\nSorted List : ", list)  # print the list
val = binary_search(list, 0, n, k)

if(val != -1):
    print("\nElement Found at index ", val)  # print index if element exists
else:
    print("\nElemetn not found!")  # print if element doesnt exists
